<a href="https://colab.research.google.com/github/SMayienda/End-Project/blob/main/KARAMOJA_PROJECT_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# KARAMOJA PROJECT

# Business Understanding

Karamoja is the most food-insecure region of Uganda. One of the main reasons is the low productivity level of the crops due to intense droughts as well as pest and disease outbreaks. Karamoja (Links to an external site.)Links to an external site.

In Karamoja, several NGOs provide technical support as well as farm inputs to the farmers experiencing extremely low yield. Though, they lack visibility into the overall state of the region and often need to rely on some very local sources of information to prioritize their activities.
Dalberg Data Insights (DDI) has been requested to develop a new food security monitoring tool to support the decision making of one of those NGOs active in Karamoja.

To do so, Dalberg Data Insights developed a methodology to remotely measure the yield of the two main staple crops of the region (i.e. sorghum and maize) based on satellite images. The agri-tech team just ran the model for the 2017 crop season.



# Data Understanding

I will be dealing with the following data:

1. Uganda Karamoja District Crop Yield
2. Uganda Karamoja Subcounty Crop Yield

Below is the data composition;

Yield and Population per Subcounty

POP: total population for the subcounty

S_Yield_Ha: average yield for sorghum for the subcounty (Kg/Ha)

M_Yield_Ha: average yield for maize for the subcounty (Kg/Ha)

Crop_Area_Ha: total crop area for the subcounty (Ha)

S_Area_Ha: total sorghum crop area for the subcounty (Ha)

M_Area_Ha: total maize crop area for the subcounty (Ha)

S_Prod_Tot: total productivity for the sorghum for the subcounty (Kg)

M_Prod_Tot: total productivity for the maize for the subcounty (Kg)

Yield and Population per District

**Research Question**

As a Data Analyst, the agri-tech team is asking you to develop an interactive visualization tool of the results for this first crop season. This visualization tool that you will develop will be used as a first mockup of the Food Security Monitoring tool that DDI will develop for the NGO.
Based on your experience, the team expects you to come up with a first draft within the coming 3 working days. They give you carte blanche in terms of structure and functionalities but they know that the client wants:

 At least a map in the dashboard

The possibility of visualizing the results by district or sub-county (two administrative levels used by the NGO)

**Objectives**

1. The population distribution across the district and their subcounties.

2. The crops productivity in each district.

3. The crops yield in each district.

4. View crops productivity verses population.

# Importing of Libraries

In [1]:
# Essential libraries to handle data manipulation and visualization.
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


# Loading of the data sets

1. District data set

In [2]:
#loading the district data
karamoja_district = pd.read_csv('/content/Uganda_Karamoja_District_Crop_Yield_Population.csv')

2. Subcounty data set

In [3]:
#loading the subcounty data
karamoja_subcounty = pd.read_csv('/content/Uganda_Karamoja_Subcounty_Crop_Yield_Population.csv')


# Data Shapes

How many columns and rows does each data set have.

In [4]:
karamoja_district.shape

(7, 11)

In [5]:
karamoja_subcounty.shape

(52, 13)

# Understand the structure and summary statistics of the data.

1. Looking into the top five raws of each data

In [6]:
karamoja_district.head()

OBJECTID     NAME     POP        Area  S_Yield_Ha  M_Yield_Ha  \
0        92     ABIM   90385  2771977106         449        1040   
1        96   AMUDAT  101790  1643582836         205        1297   
2        20  KAABONG  627057  7373606003         279         945   
3        85   KOTIDO  243157  3641539808         331        1148   
4         5   MOROTO  127811  3570160948         128         355   

   Crop_Area_Ha     S_Area_Ha    M_Area_Ha  S_Prod_Tot  M_Prod_Tot  
0   5470.068394   3277.295971  1848.621855     1471506     1922567  
1   5765.443719   2973.423860  2733.661014      609552     3545558  
2  28121.672530  20544.194960  7394.416334     5731830     6987723  
3  53032.649450  50247.443900  1751.372284    16631904     2010575  
4   5954.814048   4741.748776  1190.050606      606944      422468

In [7]:
karamoja_subcounty.head()

OBJECTID       SUBCOUNTY_NAME DISTRICT_NAME    POP        Area Karamoja  \
0       263              KACHERI        KOTIDO  17244  1067176155        Y   
1       264               KOTIDO        KOTIDO  52771   597575188        Y   
2       265  KOTIDO TOWN COUNCIL        KOTIDO  27389    23972401        Y   
3       266         NAKAPERIMORU        KOTIDO  38775   419111591        Y   
4       267           PANYANGARA        KOTIDO  65704   880955930        Y   

   S_Yield_Ha   M_Yield_Ha  Crop_Area_Ha     S_Area_Ha   M_Area_Ha  \
0  354.207411  1137.467019   7023.533691   6434.342449  528.124229   
1  367.890523  1162.996687  13587.990760  12455.592640  824.767081   
2  369.314177  1167.005832   1656.531855   1520.322052    8.561644   
3  283.324569   852.366578   7087.823334   6761.488901   45.721712   
4  373.836926  1283.859882  10398.249390  10111.198130  172.611914   

     S_Prod_Tot     M_Prod_Tot  
0  2.279092e+06  600723.892900  
1  4.582294e+06  959201.382500  
2  5.614765e+05    9991.488268  
3  1.915696e+06   38971.659080  
4  3.779939e+06  221609.511400

2. Understanding the summary of our data sets

In [8]:
karamoja_district.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7 entries, 0 to 6
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   OBJECTID      7 non-null      int64  
 1   NAME          7 non-null      object 
 2   POP           7 non-null      int64  
 3   Area          7 non-null      int64  
 4   S_Yield_Ha    7 non-null      int64  
 5   M_Yield_Ha    7 non-null      int64  
 6   Crop_Area_Ha  7 non-null      float64
 7   S_Area_Ha     7 non-null      float64
 8   M_Area_Ha     7 non-null      float64
 9   S_Prod_Tot    7 non-null      int64  
 10  M_Prod_Tot    7 non-null      int64  
dtypes: float64(3), int64(7), object(1)
memory usage: 744.0+ bytes


In [9]:
karamoja_subcounty.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52 entries, 0 to 51
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   OBJECTID        52 non-null     int64  
 1   SUBCOUNTY_NAME  52 non-null     object 
 2   DISTRICT_NAME   52 non-null     object 
 3   POP             52 non-null     int64  
 4   Area            52 non-null     int64  
 5   Karamoja        52 non-null     object 
 6   S_Yield_Ha      52 non-null     float64
 7   M_Yield_Ha      52 non-null     float64
 8   Crop_Area_Ha    52 non-null     float64
 9   S_Area_Ha       52 non-null     float64
 10  M_Area_Ha       52 non-null     float64
 11  S_Prod_Tot      52 non-null     float64
 12  M_Prod_Tot      52 non-null     float64
dtypes: float64(7), int64(3), object(3)
memory usage: 5.4+ KB


3. Understanding the statistical summary of the Data sets

In [10]:
karamoja_district.describe()

OBJECTID            POP          Area  S_Yield_Ha   M_Yield_Ha  \
count   7.000000       7.000000  7.000000e+00    7.000000     7.000000   
mean   61.714286  214943.571429  3.960853e+09  269.285714   986.142857   
std    36.481567  188604.280916  1.781860e+09  119.243049   321.566700   
min     5.000000   90385.000000  1.643583e+09  128.000000   355.000000   
25%    37.000000  114800.500000  3.171069e+09  171.000000   899.500000   
50%    80.000000  146780.000000  3.641540e+09  279.000000  1040.000000   
75%    88.500000  205391.000000  4.362553e+09  343.500000  1206.000000   
max    96.000000  627057.000000  7.373606e+09  449.000000  1297.000000   

       Crop_Area_Ha     S_Area_Ha    M_Area_Ha    S_Prod_Tot    M_Prod_Tot  
count      7.000000      7.000000     7.000000  7.000000e+00  7.000000e+00  
mean   21094.520379  16737.636651  3983.947082  4.873098e+06  4.085632e+06  
std    17363.854165  16625.963460  2678.911441  5.743724e+06  2.877188e+06  
min     5470.068394   2973.423860  1190.050606  6.069440e+05  4.224680e+05  
25%     5860.128883   4009.522373  1799.997070  1.040529e+06  1.966571e+06  
50%    22944.296020  16142.015880  2733.661014  2.211456e+06  3.545558e+06  
75%    27247.185510  19890.764085  6484.753740  6.290160e+06  6.288030e+06  
max    53032.649450  50247.443900  7394.416334  1.663190e+07  8.122197e+06

The average population is approximately 214,944, with a standard deviation of 188,604. The minimum population is 90,385, and the maximum is 627,057.

The average yield per hectare for the sorghum is 269.29, with a standard deviation of 119.24. The minimum yield is 128, and the maximum is 449

The average yield per hectare for the maize is 986.14, with a standard deviation of 321.57. The minimum yield is 355, and the maximum is 1297.

In [11]:
karamoja_subcounty.describe()

OBJECTID            POP          Area  S_Yield_Ha   M_Yield_Ha  \
count    52.000000      52.000000  5.200000e+01   52.000000    52.000000   
mean    787.865385   28934.692308  5.331913e+08  274.165405   940.259552   
std     280.101314   20865.122974  4.913308e+08  118.569907   321.641901   
min     263.000000    1418.000000  2.121209e+06  108.156411     0.000000   
25%     597.750000   16558.500000  1.568923e+08  173.034066   743.075879   
50%     810.500000   23053.500000  3.848356e+08  277.255206  1016.684002   
75%     982.250000   39461.000000  7.749029e+08  368.246437  1203.548665   
max    1320.000000  100919.000000  2.069555e+09  560.313070  1396.991494   

       Crop_Area_Ha     S_Area_Ha    M_Area_Ha    S_Prod_Tot    M_Prod_Tot  
count     52.000000     52.000000    52.000000  5.200000e+01  5.200000e+01  
mean    2839.646974   2253.143395   536.300569  6.557443e+05  5.500730e+05  
std     3110.505917   2954.355858   724.092288  9.915839e+05  7.939707e+05  
min        0.171390      0.130941     0.000000  1.728126e+01  0.000000e+00  
25%      964.876031    405.394759    79.821743  1.210555e+05  6.087012e+04  
50%     1654.265138   1231.824456   326.479336  2.543687e+05  2.896239e+05  
75%     3267.564651   2429.985069   740.296675  6.040942e+05  8.114574e+05  
max    13587.990760  12964.499730  3840.698081  4.582294e+06  4.365058e+06

The average population is approximately 28,935, with a standard deviation of 20,865. The minimum population is 1,418, and the maximum is 100,919

The average yield per hectare for the sorghum is 274.17, with a standard deviation of 118.57. The minimum yield is 108.16, and the maximum is 560.31.

The average yield per hectare for the maize is 940.26, with a standard deviation of 321.64. The minimum yield is 0, and the maximum is 1396.99.

# Missing Values and Duplicates



Checking for missing values in our data

In [12]:
karamoja_district.isnull().sum()

OBJECTID        0
NAME            0
POP             0
Area            0
S_Yield_Ha      0
M_Yield_Ha      0
Crop_Area_Ha    0
S_Area_Ha       0
M_Area_Ha       0
S_Prod_Tot      0
M_Prod_Tot      0
dtype: int64

In [13]:
karamoja_subcounty.isnull().sum()

OBJECTID          0
SUBCOUNTY_NAME    0
DISTRICT_NAME     0
POP               0
Area              0
Karamoja          0
S_Yield_Ha        0
M_Yield_Ha        0
Crop_Area_Ha      0
S_Area_Ha         0
M_Area_Ha         0
S_Prod_Tot        0
M_Prod_Tot        0
dtype: int64

In [15]:
karamoja_district.duplicated().sum()

0

In [14]:
karamoja_subcounty.duplicated().sum()

0

From the results above our data sets do not have any missing values or any duplicates.

# Combining the two data sets

In [16]:
combined_df = pd.merge(karamoja_district, karamoja_subcounty, how='outer')

<ipython-input-16-bd3d29377480>:1: UserWarning: You are merging on int and float columns where the float values are not equal to their int representation.
  combined_df = pd.merge(karamoja_district, karamoja_subcounty, how='outer')


In [17]:
# filling the missing values in the column region with mean for numeric value
numeric_columns = combined_df.select_dtypes(include=[np.number]).columns
combined_df[numeric_columns] = combined_df[numeric_columns].fillna(combined_df[numeric_columns].mean())

In [18]:
# filling missing values for non-numeric columns
non_numeric_columns = combined_df.select_dtypes(exclude=[np.number]).columns
combined_df[non_numeric_columns] = combined_df[non_numeric_columns].fillna(combined_df[non_numeric_columns].mode().iloc[0])

In [19]:
combined_df.head()

OBJECTID     NAME     POP        Area  S_Yield_Ha  M_Yield_Ha  \
0        92     ABIM   90385  2771977106       449.0      1040.0   
1        96   AMUDAT  101790  1643582836       205.0      1297.0   
2        20  KAABONG  627057  7373606003       279.0       945.0   
3        85   KOTIDO  243157  3641539808       331.0      1148.0   
4         5   MOROTO  127811  3570160948       128.0       355.0   

   Crop_Area_Ha     S_Area_Ha    M_Area_Ha  S_Prod_Tot  M_Prod_Tot  \
0   5470.068394   3277.295971  1848.621855   1471506.0   1922567.0   
1   5765.443719   2973.423860  2733.661014    609552.0   3545558.0   
2  28121.672530  20544.194960  7394.416334   5731830.0   6987723.0   
3  53032.649450  50247.443900  1751.372284  16631904.0   2010575.0   
4   5954.814048   4741.748776  1190.050606    606944.0    422468.0   

  SUBCOUNTY_NAME DISTRICT_NAME Karamoja  
0           ABIM       KAABONG        Y  
1           ABIM       KAABONG        Y  
2           ABIM       KAABONG        Y  
3           ABIM       KAABONG        Y  
4           ABIM       KAABONG        Y

In [20]:
combined_df.shape

(59, 14)

In [21]:
combined_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59 entries, 0 to 58
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   OBJECTID        59 non-null     int64  
 1   NAME            59 non-null     object 
 2   POP             59 non-null     int64  
 3   Area            59 non-null     int64  
 4   S_Yield_Ha      59 non-null     float64
 5   M_Yield_Ha      59 non-null     float64
 6   Crop_Area_Ha    59 non-null     float64
 7   S_Area_Ha       59 non-null     float64
 8   M_Area_Ha       59 non-null     float64
 9   S_Prod_Tot      59 non-null     float64
 10  M_Prod_Tot      59 non-null     float64
 11  SUBCOUNTY_NAME  59 non-null     object 
 12  DISTRICT_NAME   59 non-null     object 
 13  Karamoja        59 non-null     object 
dtypes: float64(7), int64(3), object(4)
memory usage: 6.6+ KB


In [23]:
combined_df. isnull().sum()

OBJECTID          0
NAME              0
POP               0
Area              0
S_Yield_Ha        0
M_Yield_Ha        0
Crop_Area_Ha      0
S_Area_Ha         0
M_Area_Ha         0
S_Prod_Tot        0
M_Prod_Tot        0
SUBCOUNTY_NAME    0
DISTRICT_NAME     0
Karamoja          0
dtype: int64

In [25]:
combined_df.describe()

OBJECTID            POP          Area  S_Yield_Ha   M_Yield_Ha  \
count    59.000000      59.000000  5.900000e+01   59.000000    59.000000   
mean    701.711864   51003.542373  9.398630e+08  273.586459   945.703334   
std     353.856198   87994.178609  1.338079e+09  117.624497   319.200199   
min       5.000000    1418.000000  2.121209e+06  108.156411     0.000000   
25%     592.500000   16721.000000  1.901488e+08  170.240064   757.082394   
50%     767.000000   26644.000000  4.997769e+08  279.000000  1030.064093   
75%     980.500000   43618.000000  1.048900e+09  361.945262  1211.009291   
max    1320.000000  627057.000000  7.373606e+09  560.313070  1396.991494   

       Crop_Area_Ha     S_Area_Ha    M_Area_Ha    S_Prod_Tot    M_Prod_Tot  
count     59.000000     59.000000    59.000000  5.900000e+01  5.900000e+01  
mean    5005.479412   3971.642595   945.343375  1.156108e+06  9.695461e+05  
std     8668.593760   7654.205284  1570.917869  2.483804e+06  1.655393e+06  
min        0.171390      0.130941     0.000000  1.728126e+01  0.000000e+00  
25%      998.266305    533.066508    92.508396  1.466574e+05  7.862814e+04  
50%     2008.068169   1550.944570   358.550335  3.397606e+05  3.069518e+05  
75%     5616.531290   3216.917338   957.467889  9.664307e+05  9.700422e+05  
max    53032.649450  50247.443900  7394.416334  1.663190e+07  8.122197e+06

In [26]:
from google.colab import files
combined_df.to_csv('combined_df.csv')
files.download('combined_df.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

# Conclusion

1. The population for both district and subcounty is evenly distributed except for Kaabong whose population is very high.

2. Kaabong has the highest population in district and subcounty level with low crop productivity.

3. Nakapiripirit seems to have a balance between population and crop productivity and crop yield.

4. Moronto has the lowest productivity which does not relate to its population which is higher.


# Recommendation

1. More support needs to be given to Kaabong district to improve its productivity which will match its population that is very high.

2. Support is also needed in Moroto to improve its productivity.

3. The same applies to Amudata which has very low sorghum production.

4. The farmers should try crop diversification.

# Future Work

In the future i would like to work with this data set having included the following information.
1. Soil type in each region.
2. Climate seasons.
3. The seed and fertelizer used is common across all the district.